src/tsolvers/axdiffsolver/ArraySolver.cc

```diff
@@ -0,0 +1,788 @@
+    auto steps2 = cursor2.countSecondaryEdges(index);
+    while (steps1 > steps2) {
+        cursor1.collectOneSecondary(indices, index);
+        --steps1;
+    }
+    while (steps2 > steps1) {
+        cursor2.collectOneSecondary(indices, index);
+        --steps2;
+    }
+    while (findSecondaryNode(cursor1.getCurrentNodeRef(), index) != findSecondaryNode(cursor2.getCurrentNodeRef(), index)) {
+        cursor1.collectOneSecondary(indices, index);
+        cursor2.collectOneSecondary(indices, index);
+    }
+    cursor1.collectOverPrimaries(indices, cursor2.getCurrentNodeRef());
+    return indices;
+}
+
+void ArraySolver::computeExplanation(PTRef equality) {
+    auto conflictExplanation = readOverWeakEquivalenceConflict(equality);
+    has_explanation = true;
+    explanation.clear();
+    for (auto lit : conflictExplanation) {
+        explanation.push(lit);
+    }
+}
+
+/*
+ * Compute the literals representing the negation of a read-over-weak-eq lemma, i.e., the conflict, for the given
+ * equality in the current context.
+ */
+ArraySolver::ExplanationCollection ArraySolver::readOverWeakEquivalenceConflict(PTRef equality) {
+    assert(logic.isEquality(equality));
+    PTRef lhs = logic.getPterm(equality)[0];
+    PTRef rhs = logic.getPterm(equality)[1];
+    assert(logic.isArraySelect(logic.getSymRef(lhs)));
+    assert(logic.isArraySelect(logic.getSymRef(rhs)));
+    ERef array1 = egraph.termToERef(logic.getPterm(lhs)[0]);
+    ERef array2 = egraph.termToERef(logic.getPterm(rhs)[0]);
+    ERef index1 = egraph.termToERef(logic.getPterm(lhs)[1]);
+    ERef index2 = egraph.termToERef(logic.getPterm(rhs)[1]);
+
+    // collect literals explaining why array1 is weakly equivalent to array2
+    auto lemma = explainWeakEquivalencePath(array1, array2, getRoot(index1));
+    // collect literals explaining why index1 is equivalent to index2 in Egraph
+    if (index1 != index2) {
+        recordExplanationOfEgraphEquivalence(lemma, index1, index2);
+    }
+    // add literal asserting that the selects are not equal
+    lemma.insert({equality, l_False});
+    return lemma;
+}
+
+/*
+ * Explain why the two input arrays are weakly equivalent on index "index".
+ * Since they are weakly equivalent on "index", the selects "array1[index]" and "array2[index]" must have the same value.
+ *
+ * @returns The collection of literals that guarantees the i-weak equivalence
+ */
+ArraySolver::ExplanationCollection ArraySolver::explainWeakEquivalencePath(ERef array1, ERef array2, ERef index) {
+    assert(getRoot(index) == index);
+    NodeRef node1 = getNodeRef(getRoot(array1));
+    NodeRef node2 = getNodeRef(getRoot(array2));
+    assert(getIndexedRepresentative(node1, index) == getIndexedRepresentative(node2, index));
+
+    Traversal traversal(*this);
+    IndicesCollection storeIndices;
+    ExplanationCollection explanations;
+
+    auto count1 = traversal.countSecondaryEdges(node1, index);
+    auto count2 = traversal.countSecondaryEdges(node2, index);
```

Comment:
Most likely it should work.
I basically followed SMTInterpol implementation, as I did not have to think about it that much.

We can try to change it if this would be a performance bottleneck, but I somewhat doubt that is the case.

Reply:
Ok. In any case I would push the action point to the future.